File: backend/core-service/src/main/kotlin/ai/govbiz/core/admin/domain/AdminAccount.kt
package ai.govbiz.core.admin.domain

import ai.govbiz.core.account.domain.AccountRole
import ai.govbiz.core.account.domain.AccountTier
import ai.govbiz.core.account.domain.OAuthProvider
import java.time.LocalDateTime

/** 관리자 목록에서 계정을 나누는 상태입니다. 삭제된 계정은 목록과 상세에 나오지 않습니다. */
enum class AdminAccountStatus {
    ACTIVE,
    SUSPENDED,
}

/** 로그인 방법입니다. 이메일은 비밀번호가 있는 계정, 카카오·Google은 소셜 로그인이 연결된 계정이며 한 계정이 여럿을 가질 수 있습니다. */
enum class AdminAccountLoginMethod {
    EMAIL,
    KAKAO,
    GOOGLE,
}

/** 목록 정렬입니다. 최근 로그인순은 로그인한 적 없는 계정을 뒤에 둡니다. */
enum class AdminAccountSort {
    RECENT,
    OLDEST,
    LAST_LOGIN,
}

/** 관리자 조치 종류입니다. DB CHECK 제약과 같은 값입니다. */
enum class AdminAccountActionType {
    SUSPEND,
    UNSUSPEND,
    SESSIONS_REVOKE,
}

/** 관리자 계정 목록을 좁히는 조건입니다. 상태·역할·로그인 방법이 null이면 전체입니다. */
data class AdminAccountQuery(
    /** 이메일·기업명, 숫자만 있으면 사업자등록번호 일부로도 찾습니다. */
    val keyword: String,
    val status: AdminAccountStatus?,
    val role: AccountRole?,
    val loginMethod: AdminAccountLoginMethod?,
    val sort: AdminAccountSort,
    val page: Int,
    val pageSize: Int,
) {
    init {
        require(keyword == keyword.trim() && keyword.length <= MAX_KEYWORD_LENGTH) {
            "keyword must be a trimmed text of at most $MAX_KEYWORD_LENGTH characters"
        }
        require(page >= 1) { "page must be positive" }
        require(pageSize in 1..MAX_PAGE_SIZE) { "pageSize must be 1~$MAX_PAGE_SIZE" }
    }

    val offset: Int
        get() = (page - 1) * pageSize

    companion object {
        const val MAX_KEYWORD_LENGTH = 100
        const val MAX_PAGE_SIZE = 50
    }
}

/** 관리자 목록 한 줄입니다. 비밀번호 해시·세션 토큰은 담지 않습니다. */
data class AdminAccountSummary(
    val id: Long,
    val email: String,
    val role: AccountRole,
    val emailVerified: Boolean,
    val hasPassword: Boolean,
    val oauthProviders: Set<OAuthProvider>,
    val companyName: String?,
    val businessNumber: String?,
    val suspendedAt: LocalDateTime?,
    val createdAt: LocalDateTime,
    /** 로그인·가입·소셜 로그인으로 세션을 받은 마지막 시각입니다. 이 값이 생기기 전의 로그인은 남아 있지 않아 null입니다. */
    val lastLoginAt: LocalDateTime?,
) {
    val status: AdminAccountStatus
        get() = if (suspendedAt == null) AdminAccountStatus.ACTIVE else AdminAccountStatus.SUSPENDED

    /** 회원 화면의 권한 단계와 같은 규칙입니다. */
    val tier: AccountTier
        get() = when {
            role == AccountRole.ADMIN -> AccountTier.ADMIN
            companyName != null -> AccountTier.COMPANY
            else -> AccountTier.MEMBER
        }

    val loginMethods: List<AdminAccountLoginMethod>
        get() = buildList {
            if (hasPassword) add(AdminAccountLoginMethod.EMAIL)
            if (OAuthProvider.KAKAO in oauthProviders) add(AdminAccountLoginMethod.KAKAO)
            if (OAuthProvider.GOOGLE in oauthProviders) add(AdminAccountLoginMethod.GOOGLE)
        }
}

/** 한 페이지 결과입니다. 총 건수는 같은 조건의 전체 건수입니다. */
data class AdminAccountPage(
    val accounts: List<AdminAccountSummary>,
    val total: Long,
    val page: Int,
    val pageSize: Int,
) {
    val totalPages: Int
        get() = if (total == 0L) 0 else ((total - 1) / pageSize + 1).toInt()
}

/** 목록 위 요약 수치입니다. 삭제된 계정은 세지 않습니다. */
data class AdminAccountStats(
    val total: Long,
    val companyRegistered: Long,
    val socialLinked: Long,
    val suspended: Long,
    val admins: Long,
    /** 최근 [ai.govbiz.core.admin.service.AdminAccountService.RECENT_JOIN_DAYS]일 안에 가입한 계정 수입니다. */
    val joinedRecently: Long,
)

/** 상세의 기업 정보입니다. */
data class AdminAccountCompany(
    val companyName: String,
    val businessNumber: String,
    val region: String,
    val industry: String,
    val foundedYear: Int,
)

/** 상세의 활동 수치입니다. 세션은 아직 만료되지 않은 것만 셉니다. */
data class AdminAccountActivity(
    val recruitmentCount: Int,
    val openRecruitmentCount: Int,
    val sentProposalCount: Int,
    val activeSessionCount: Int,
)

/** 관리자 조치 기록 한 건입니다. */
data class AdminAccountAction(
    val id: Long,
    val action: AdminAccountActionType,
    val reason: String,
    val adminEmail: String,
    val createdAt: LocalDateTime,
)

data class AdminAccountDetail(
    val account: AdminAccountSummary,
    val company: AdminAccountCompany?,
    val activity: AdminAccountActivity,
    /** 최근 조치부터 최대 [ai.govbiz.core.admin.service.AdminAccountService.ACTION_HISTORY_LIMIT]건입니다. */
    val actions: List<AdminAccountAction>,
)

/** 조치 transaction 안에서 잠가 읽은 대상 계정의 상태입니다. */
data class AdminAccountTarget(
    val id: Long,
    val role: AccountRole,
    val suspendedAt: LocalDateTime?,
)

File: frontend/packages/shared/src/domain/entities/AdminAccount.ts
import type { AccountRole, AccountTier } from './Account'

/** 관리자 목록에서 계정을 나누는 상태입니다. 삭제된 계정은 목록과 상세에 나오지 않습니다. */
export type AdminAccountStatus = 'ACTIVE' | 'SUSPENDED'

/** 로그인 방법입니다. 이메일은 비밀번호가 있는 계정이며 한 계정이 여러 방법을 가질 수 있습니다. */
export type AdminAccountLoginMethod = 'EMAIL' | 'KAKAO' | 'GOOGLE'

/** 목록 정렬입니다. 최근 로그인순은 로그인한 적 없는 계정을 뒤에 둡니다. */
export type AdminAccountSort = 'RECENT' | 'OLDEST' | 'LAST_LOGIN'

/** 관리자 조치 기록의 종류입니다. */
export type AdminAccountActionType = 'SUSPEND' | 'UNSUSPEND' | 'SESSIONS_REVOKE'

/** 화면에서 고르는 조치입니다. 서버 경로와 기록 종류로 바꾸는 일은 Data Layer가 맡습니다. */
export type AdminAccountActionKind = 'suspend' | 'unsuspend' | 'revoke-sessions'

/** 관리자 목록 한 줄입니다. 비밀번호·토큰은 없고 비밀번호가 있는지만 압니다. 시각은 서울 기준 `yyyy-MM-ddTHH:mm:ss`입니다. */
export type AdminAccountSummary = {
  id: number
  email: string
  role: AccountRole
  tier: AccountTier
  status: AdminAccountStatus
  emailVerified: boolean
  hasPassword: boolean
  loginMethods: AdminAccountLoginMethod[]
  company: { companyName: string; businessNumber: string } | null
  createdAt: string
  lastLoginAt: string | null
  suspendedAt: string | null
}

/** 목록 조건입니다. 상태·역할·로그인 방법이 빈 문자열이면 전체입니다. */
export type AdminAccountQuery = {
  keyword: string
  status: AdminAccountStatus | ''
  role: AccountRole | ''
  loginMethod: AdminAccountLoginMethod | ''
  sort: AdminAccountSort
  page: number
}

export const adminAccountPageSize = 20

export const defaultAdminAccountQuery: AdminAccountQuery = {
  keyword: '',
  status: '',
  role: '',
  loginMethod: '',
  sort: 'RECENT',
  page: 1,
}

export type AdminAccountPage = {
  accounts: AdminAccountSummary[]
  total: number
  page: number
  pageSize: number
  totalPages: number
}

/** 목록 위 요약 수치입니다. `joinedRecently`는 최근 `recentJoinDays`일 안에 가입한 계정 수입니다. */
export type AdminAccountStats = {
  total: number
  companyRegistered: number
  socialLinked: number
  suspended: number
  admins: number
  joinedRecently: number
  recentJoinDays: number
}

export type AdminAccountAction = {
  id: number
  action: AdminAccountActionType
  reason: string
  adminEmail: string
  createdAt: string
}

export type AdminAccountDetail = {
  account: AdminAccountSummary
  company: {
    companyName: string
    businessNumber: string
    region: string
    industry: string
    foundedYear: number
  } | null
  activity: {
    recruitmentCount: number
    openRecruitmentCount: number
    sentProposalCount: number
    activeSessionCount: number
  }
  /** 최근 조치부터 최대 20건입니다. */
  actions: AdminAccountAction[]
  /** 조회한 관리자 자신의 계정이면 참입니다. */
  isSelf: boolean
}

/** 조치 사유는 서버와 같이 앞뒤 공백을 뺀 1~500자입니다. */
export const adminActionReasonMaxLength = 500

export const adminAccountStatusLabels: Record<AdminAccountStatus, string> = {
  ACTIVE: '정상',
  SUSPENDED: '정지',
}

export const adminAccountRoleLabels: Record<AccountRole, string> = {
  USER: '회원',
  ADMIN: '관리자',
}

export const adminAccountLoginMethodLabels: Record<AdminAccountLoginMethod, string> = {
  EMAIL: '이메일',
  KAKAO: '카카오',
  GOOGLE: 'Google',
}

export const adminAccountSortLabels: Record<AdminAccountSort, string> = {
  RECENT: '최근 가입순',
  OLDEST: '오래된 가입순',
  LAST_LOGIN: '최근 로그인순',
}

export const adminAccountActionLabels: Record<AdminAccountActionType, string> = {
  SUSPEND: '정지',
  UNSUSPEND: '정지 해제',
  SESSIONS_REVOKE: '강제 로그아웃',
}

/** 조치할 수 있는 계정인지입니다. 자기 계정과 다른 관리자 계정에는 서버도 조치를 거절합니다. */
export function availableAdminAccountActions(detail: AdminAccountDetail): AdminAccountActionKind[] {
  if (detail.isSelf || detail.account.role === 'ADMIN') return []
  return detail.account.status === 'SUSPENDED' ? ['unsuspend'] : ['suspend', 'revoke-sessions']
}
